libp2p/NodeTable.cpp

```diff
@@ -233,22 +234,17 @@ void NodeTable::doDiscover(NodeID _node, unsigned _round, shared_ptr<set<shared_
         return;
     }
 
-    m_timers.schedule(c_reqTimeout.count() * 2, [this, _node, _round, _tried](boost::system::error_code const& _ec)
-    {
+    m_discoveryTimer.expires_from_now(boost::posix_time::milliseconds(c_reqTimeout.count() * 2));
+    m_discoveryTimer.async_wait([this, _node, _round, _tried](
+                                    boost::system::error_code const& _ec) {
         if (_ec)
+        {
```

Comment:
Only checking `_ec` is not enough. This will handle only the case where the timer was cancelled before it expired. But there is also the case where the timer expired and handler in already in the completion queue. In this case there is not error code. So `doDiscover()` will be called again and new timer scheduled...

To handle this case there are two requirements:
- The NodeTable must still exist, what's why using shared_ptr is handy here.
- We have to check if NodeTable is in valid state before doing `doDiscover()`. We can use any `m_running` flag, but a smart way is to set the timer's `expires_at` to some arbitrary value like `time_point::min()`.

See https://stackoverflow.com/a/43169596/725174.

Reply:
Quick fix would be to handle timers by shared_ptrs, i.e. 
```
std::shared_ptr<ba::deadline_timer> m_discoveryTimer;
```
and pass it also to the handler so it can access the timer safely and check if `expires_at` is not the `min` value.

Reply:
> See https://stackoverflow.com/a/43169596/725174.

Thanks, that's an awesome link 😄 I didn't realize that calling `cancel `after a timer's handler has already been queued wouldn't affect its error code.

I'll make the `time_point::min` + `shared_ptr` change, though I don't think they're technically needed for a couple of reasons:
*  The node table lifetime (and therefore the timer lifetimes) is tied to the host's lifetime, and when the host is destroyed then there's no more ioservice work loop meaning that the node table is guaranteed to exist if the timer handlers are executing.
* I don't think `doDiscover` would be called if its handler is executed after timer cancellation because we cancel the timers in `NodeTable::stop()` and we also close the UDP socket, and the first thing that `doDiscover` does is check if the UDP socket is still open...so `doDiscover ` should immediately exit if it executes after the timers have been cancelled. 

The eviction timer handler however would still execute the eviction code...I took a quick look at it and I don't think we'd hit any major issues as long as the node table hasn't been destroyed but I'm not completely sure.

I do think it's good practice to keep the timers alive through `shared_ptr `rather than relying on the timers' owner's lifetime though (since we may refactor things later).

Reply:
> 
> 
> This looks good to me, but there are some test failures that must be investigated.
> 
> Moreover, check out this tutorial about triggering timers periodically (we are close to this, but maybe you can spot some small differences): https://www.boost.org/doc/libs/1_69_0/doc/html/boost_asio/tutorial/tuttimer3.html
> 
> Also they are using `steady_timer` instead of `deadline_timer`. The difference is only in the clock used. We don't need high precision here, so maybe one is better than the other for our case.

steady_clock might be nice since we'd be able to switch to std::chrono for timers and it wouldn't be impacted like deadline timers if someone changes the system clock while Aleth is running. However, I need to read a little more about its precision before deciding if we should move over.

Reply:
Steady-clock appears to provide 1/100 microsecond precision on both Windows and Linux so I think we can use it, I'd like to make this change in a future PR though.